neps/nep-0413.md

```diff
@@ -0,0 +1,142 @@
+
+### How to Return the Signed Message in a Safe Way
+Sending the signed message in a query string to an arbitrary URL (even within the correct domain) is not secure as the data can be leaked (e.g. through headers, etc). Using URL fragments instead will improve security, since [URL fragments are not included in the `Referer`](https://greenbytes.de/tech/webdav/rfc2616.html#header.referer).
+
+### NEAR Signatures
+NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+## Specification
+Wallets must implement a `signMessage` method, which takes a `message` destined to a specific `receiver` and transform it into a verifiable signature.
+
+### Input Interface
+`signMessage` must implement the following input interface:
+
+```jsx
+interface SignMessageParams {
+  message: string ; // The message that wants to be transmitted.
+  receiver: string; // The receiver to whom the message is destined (e.g. "alice.near" or "myapp.com").
+  nonce: [u8; 32] ; // A nonce that uniquely identifies this instance of the message, denoted as a 32 bytes array (a fixed `Buffer` in JS/TS).
+  callbackUrl?: string; // Optional, applicable to browser wallets (e.g. MyNearWallet). The URL to call after the signing process. Defaults to `window.location.href`.
+}
+```
+
+### Structure
+`signMessage` must embed the input `message`, `receiver` and `nonce` into the following predefined structure:
+
+```rust
+struct Payload {
+  message: string; // The same message passed in `SignMessageParams.message`
+  receiver: string; // The same receiver passed in `SignMessageParams.receiver`
+  nonce: [u8; 32]; // The same nonce passed in `SignMessageParams.nonce`
+}
+```
+
+### Signature
+In order to create a signature, `signMessage` must:
+1. Convert the `Payload` into its [JSON JCS](https://www.rfc-editor.org/rfc/rfc8785) string representation (i.e. a JSON string, with alphabetically ordered attributes).
+2. Prepend the `NEP0413:` string to the result from step 1.
+3. Compute the `SHA256` hash of the result from step 2.
+4. Sign the resulting `SHA256` hash from step 3 using a **full-access** key.
+
+> If the wallet does not hold any `full-access` keys, then it must return an error.
+
+### Example
+Assuming that the `signMessage` method was invoked, and that:
+- The input `message` is `"hi"`
+- The input `receiver` is `"myapp.com"`
+- The input `nonce` is `[0,...,31]`
+- The wallet stores a full-access private key
+
+The wallet must construct and sign the following `SHA256` hash:
+
+```jsx
+sha256.hash(`NEP0413:` + `{"message":"hi","receiver":"myapp.com","nonce":"[0,...,31]"}`)
```

Comment:
I believe the keys should be sorted alphabetically and the nonce should be an array without quotes.

```
sha256.hash(`NEP0413:` + `{"message":"hi","nonce":[0,...,31],"receiver":"myapp.com"}`)
```

Reply:
Indeed, the keys must be sorted, as the NEP currently specifies: "(i.e. a JSON string, with alphabetically ordered attributes)".

I'll fix the quotes around the array, those are a mistake in the example. Thanks for spoting it!

Reply:
@gagdiez can you also move the `nonce` attribute before `receiver` so that the keys are sorted (as per my snippet above)?

Reply:
Fixed, thank you!